### PR TITLE
Fix save-last behavior

### PR DESCRIPTION
**Pull request by zphang**
This commit makes explicit separate save-best and save-last behaviors, into `best_model.p` and (new) `last_model.p`.

### CODE REVIEW

File: jiant/proj/main/metarunner.py [hunk 1]
```diff
@@ -122,6 +125,27 @@ def save_model(self):
...
+        save_state_dict_with_metadata(
+            state_dict=self.best_state_dict,
+            metadata={
+                "val_state": self.best_val_state.to_dict(),
+                "val_metrics": self.best_val_state.metrics,
+            },
+            output_dir=self.output_dir,
+            file_name="best_model",
+        )
+
```
Comment (jeswan) [inline jiant/proj/main/metarunner.py:148]: These functions seem like they are doing the same thing in different ways. Can they be combined?

Comment (jeswan) [reply jiant/proj/main/metarunner.py:148]: This also looks like a potential case of excessive delegation.

**Comment by jeswan** [comment]
Please add a test for these additions or create an issue for a test and assign it to me.

**Review by jeswan — CHANGES_REQUESTED** [review]
Please add a test for these additions or create an issue for a test and assign it to me.

File: jiant/proj/main/metarunner.py (hunk 1 above)
Comment (zphang) [reply jiant/proj/main/metarunner.py:148]: It looks excessive, but the specific reason for this is to allow these methods to be overridden. (E.g. for adapters, we don't want to save all the model's weights, so we need a custom save function). Happy to implement another way, but I've found this helpful.

Comment (zphang) [reply jiant/proj/main/metarunner.py:148]: Side benefit: can override for testing

Comment (jeswan) [reply jiant/proj/main/metarunner.py]: I still think there is room for refactoring here. The function signature should similar to:
```python
def save_model(self, model, file_name, metadata_dict)
```

File: jiant/shared/runner.py [hunk 2]
```diff
@@ -55,20 +56,16 @@ def get_eval_dataloader_from_cache(
...
+        state_dict = model_or_state_dict
+    else:
+        state_dict = torch_utils.get_model_for_saving(model_or_state_dict).state_dict()
 
-def save_state_dict_with_metadata(
-    state_dict: dict, metadata: dict, output_dir: str, file_name="model"
-):
     torch.save(state_dict, os.path.join(output_dir, f"{file_name}.p"))
-    py_io.write_json(metadata, os.path.join(output_dir, f"{file_name}.metadata.json"))
+    if metadata is not None:
```
Comment (jeswan) [inline jiant/shared/runner.py:71]:
```suggestion
    if metadata:
```

File: jiant/shared/runner.py [hunk 3]
```diff
@@ -55,20 +56,16 @@ def get_eval_dataloader_from_cache(
...
+        py_io.write_json(metadata, os.path.join(output_dir, f"{file_name}.metadata.json"))
```
Comment (jeswan) [inline jiant/shared/runner.py:72]:
```suggestion
    if metadata:
        py_io.write_json(metadata, os.path.join(output_dir, f"{file_name}.metadata.json"))
```

File: jiant/shared/runner.py [hunk 4]
```diff
@@ -55,12 +56,20 @@ def get_eval_dataloader_from_cache(
...
 
 
-def save_model_with_metadata(model: nn.Module, metadata: dict, output_dir: str, file_name="model"):
-    torch.save(
-        torch_utils.get_model_for_saving(model).state_dict(),
-        os.path.join(output_dir, f"{file_name}.p"),
-    )
-    py_io.write_json(metadata, os.path.join(output_dir, f"{file_name}.metadata.json"))
+def save_model_with_metadata(
+    model_or_state_dict: Union[nn.Module, dict],
```
Comment (jeswan) [inline jiant/shared/runner.py:60]:
```suggestion
    state_dict: dict,
```

File: jiant/shared/runner.py [hunk 5]
```diff
@@ -55,12 +56,20 @@ def get_eval_dataloader_from_cache(
...
+    output_dir: str,
+    file_name="model",
+    metadata: Optional[dict] = None,
+):
+    if isinstance(model_or_state_dict, dict):
+        state_dict = model_or_state_dict
+    else:
+        state_dict = torch_utils.get_model_for_saving(model_or_state_dict).state_dict()
```
Comment (jeswan) [inline jiant/shared/runner.py:68]:
```suggestion
```

File: jiant/shared/runner.py (hunk 4 above)
Comment (jeswan) [inline jiant/shared/runner.py:60]:
```suggestion
def save_state_dict_metadata(
    state_dict: dict,
```

File: jiant/proj/main/metarunner.py [hunk 6]
```diff
@@ -218,20 +239,12 @@ def eval_save(self):
...
-                    },
-                    output_dir=self.output_dir,
-                    file_name="best_model",
-                )
             del self.best_state_dict
             self.best_state_dict = copy_state_dict(
                 state_dict=get_model_for_saving(self.model).state_dict(), target_device=CPU_DEVICE,
             )
+            if self.save_best_model:
+                self.save_best_model_with_metadata(val_metrics_dict=val_metrics_dict)
```
Comment (jeswan) [inline jiant/proj/main/metarunner.py:247]:
```suggestion
                self.save_state_dict(state_dict=self.best_state_dict, output_dir=self.output_dir, metadata=val_metrics_dict, file_name="best_model")
```

File: jiant/shared/runner.py [hunk 7]
```diff
@@ -55,12 +56,20 @@ def get_eval_dataloader_from_cache(
...
-    torch.save(
-        torch_utils.get_model_for_saving(model).state_dict(),
-        os.path.join(output_dir, f"{file_name}.p"),
-    )
-    py_io.write_json(metadata, os.path.join(output_dir, f"{file_name}.metadata.json"))
+def save_model_with_metadata(
+    model_or_state_dict: Union[nn.Module, dict],
+    output_dir: str,
+    file_name="model",
+    metadata: Optional[dict] = None,
```
Comment (jeswan) [inline jiant/shared/runner.py:63]:
```suggestion
    model: Optional,
    file_name="model",
    metadata: Optional[dict] = None,
```

**Comment by jeswan** [comment]
Planning to present a refactored alternative to `save_state_dict_with_metadata` is conflated with `save_model_with_metadata` in runner. There is not a standard format for saving a model vs state dict currently there is excessive delegation ie: `save_model_with_metadata` in runner.py calls `save_state_dict_with_metadata`.

**Comment by zphang** [comment]
Mind taking a look at the changes again? I think the updates addressed some of your comments.

**Review by jeswan — APPROVED** [review]
Thanks for addressing my comments! Looks cleaner!

**Comment by HarshTrivedi** [comment]
@zphang @jeswan It seems that `self.save_last_model` flag isn't being used anywhere. The done_training method *always* calls this method: `save_last_model_with_metadata`.

**Comment by zphang** [comment]
Thanks! https://github.com/nyu-mll/jiant/pull/1243

**Pull request merged**